src/ragas/metrics/context_relevance.py

```diff
@@ -2,62 +2,137 @@
-        self.model = QGen(self.model_name, self.device)
-
-    @staticmethod
-    def _make_question_context_rank_pairs(row: dict) -> dict:
-        q_id = row["q_id"]
-        q = row["question"]
-        cs = row["contexts"]
-        new_q_ids = []
-        sentences = []
-        for i, q_id in enumerate(q_id):
-            for c in cs[i]:
-                sentences.append([q[i], c])
-                new_q_ids.append(q_id)
-
-        return {"q_id": new_q_ids, "sentences": sentences}
+        self.sent_agreement = SentenceAgreement(
+            model_name=self.model_name, metric=self.agreement_metric
+        )
 
     def score(self: t.Self, dataset: Dataset) -> Dataset:
         """
         dataset: Dataset[question: list[str], contexts: list[list[str]]]
         """
+        prompts = []
+        questions, contexts = dataset["question"], dataset["contexts"]
+        for q, c in zip(questions, contexts):
+            prompt = CONTEXT_RELEVANCE.format(q, "\n".join(c))
+            prompts.append(prompt)
 
-        num_qs, _ = dataset.shape
-        dataset = dataset.add_column("q_id", list(range(num_qs)))  # type: ignore
-        sentence_ds = dataset.map(
-            self._make_question_context_rank_pairs,
-            batched=True,
-            batch_size=10,
-            remove_columns=dataset.column_names,
-        )
+        responses = openai_completion(prompts, n=self.strictness)
+
+        prev = 0
+        outputs = []
+        responses = responses["choices"]  # type: ignore
+        for i in range(self.strictness, len(responses) + 1, self.strictness):
+            output = [responses[idx]["text"].strip() for idx in range(prev, i)]
 
-        # we loose memory here because we have to make it py_list
-        scores = self.model.predict(sentence_ds["sentences"])
+            outputs.append(output)
+            prev = i
 
-        sentence_ds = sentence_ds.add_column(self.name, scores)  # type: ignore
-        df = sentence_ds.to_pandas()
-        assert isinstance(df, DataFrame)
-        score_ds = Dataset.from_dict({self.name: df.groupby("q_id")[self.name].mean()})
+        scores = []
+        for context, n_output in zip(contexts, outputs):
+            context = "\n".join(context)
+            overlap_scores = []
+            context_sents = sent_tokenize(context)
+            for output in n_output:
+                indices = [
+                    context.find(sent)
+                    for sent in sent_tokenize(output)
+                    if context.find(sent) != -1
+                ]
+                overlap_scores.append(len(indices) / len(context_sents))
+            if self.strictness > 1:
+                agr_score = self.sent_agreement.evaluate(n_output)
+            else:
+                agr_score = 1
+            scores.append(agr_score * np.mean(overlap_scores))
 
-        return score_ds
+        return dataset.add_column(f"{self.name}", scores)  # type: ignore
```

Comment:
why are we type ignoring here? just curious

Reply:
A warning for not passing the dataset fingerprint argument, The same is done in the Faithfulness class.